src/btree/bt_handle.c

```diff
@@ -688,18 +688,21 @@ __btree_page_sizes(WT_SESSION_IMPL *session)
 
 	/*
 	 * Don't let pages grow large compared to the cache size or we can end
-	 * up in a situation where nothing can be evicted.  Take care getting
-	 * the cache size: with a shared cache, it may not have been set.
-	 * Don't forget to update the API documentation if you alter the
-	 * bounds for any of the parameters here.
+	 * up in a situation where nothing can be evicted.  Make sure at least
+	 * 10 pages fit in cache when it is at the dirty trigger where threads
+	 * stall.
+	 *
+	 * Take care getting the cache size: with a shared cache, it may not
+	 * have been set.  Don't forget to update the API documentation if you
+	 * alter the bounds for any of the parameters here.
 	 */
 	WT_RET(__wt_config_gets(session, cfg, "memory_page_max", &cval));
 	btree->maxmempage = (uint64_t)cval.val;
-	if (!F_ISSET(conn, WT_CONN_CACHE_POOL)) {
-		if ((cache_size = conn->cache_size) > 0)
-			btree->maxmempage =
-			    WT_MIN(btree->maxmempage, cache_size / 10);
-	}
+	if (!F_ISSET(conn, WT_CONN_CACHE_POOL) &&
+	    (cache_size = conn->cache_size) > 0)
+		btree->maxmempage = WT_MIN(btree->maxmempage,
+		    (conn->cache->eviction_dirty_trigger * cache_size) / 1000);
```

Comment:
I'm going to merge, but a more detailed comment here about how the computation above, dividing by 1000 and adding the multiplication of dirty trigger yields 10 pages would be helpful.